index/segment/tests/fst_mem_simple_test.go

```diff
@@ -0,0 +1,283 @@
+}
+
+func TestTermEqualsWithFewDocs(t *testing.T) {
+	testTermsEqual(t, fewTestDocuments)
+}
+
+func TestTermEqualsWithLotsDocs(t *testing.T) {
+	testTermsEqual(t, lotsTestDocuments)
+}
+
+func TestPostingsListEqualForMatchTermFewDocs(t *testing.T) {
+	testPostingsListEqualForMatchTerm(t, fewTestDocuments)
+}
+
+func TestPostingsListEqualForMatchTermLotsDocs(t *testing.T) {
+	testPostingsListEqualForMatchTerm(t, lotsTestDocuments)
+}
+
+func testPostingsListEqualForMatchTerm(t *testing.T, docs []doc.Document) {
+	memSeg, fstSeg := newTestSegments(t, docs)
+
+	memFields, err := memSeg.Fields()
+	require.NoError(t, err)
+
+	for _, f := range memFields {
+		memTerms, err := memSeg.Terms(f)
+		require.NoError(t, err)
+
+		for _, term := range memTerms {
+			memPl, err := memSeg.MatchTerm(f, term)
+			require.NoError(t, err)
+			fstPl, err := fstSeg.MatchTerm(f, term)
+			require.NoError(t, err)
+			require.True(t, memPl.Equal(fstPl),
+				fmt.Sprintf("%s:%s - [%v] != [%v]", string(f), string(term), pprintIter(memPl), pprintIter(fstPl)))
+		}
+	}
+}
+
+func pprintIter(pl postings.List) string {
+	var buf bytes.Buffer
+	iter := pl.Iterator()
+	for i := 0; iter.Next(); i++ {
+		if i != 0 {
+			buf.WriteString(", ")
+		}
+		buf.WriteString(fmt.Sprintf("%d", iter.Current()))
+	}
+	return buf.String()
+}
+
+func testTermsEqual(t *testing.T, docs []doc.Document) {
+	memSeg, fstSeg := newTestSegments(t, docs)
+
+	memFields, err := memSeg.Fields()
+	require.NoError(t, err)
+	fstFields, err := fstSeg.Fields()
+	require.NoError(t, err)
+
+	assertTermEquals := func(fields [][]byte) {
+		for _, f := range fields {
+			memTerms, err := memSeg.Terms(f)
+			require.NoError(t, err)
+			fstTerms, err := fstSeg.Terms(f)
+			require.NoError(t, err)
+			assertSliceOfByteSlicesEqual(t, memTerms, fstTerms)
+		}
+	}
+	assertTermEquals(memFields)
+	assertTermEquals(fstFields)
+}
+
+func testFieldsEquals(t *testing.T, docs []doc.Document) {
+	memSeg, fstSeg := newTestSegments(t, docs)
+
+	memFields, err := memSeg.Fields()
+	require.NoError(t, err)
+
+	fstFields, err := fstSeg.Fields()
+	require.NoError(t, err)
+
+	assertSliceOfByteSlicesEqual(t, memFields, fstFields)
+}
+
+func assertSliceOfByteSlicesEqual(t *testing.T, a, b [][]byte) {
+	require.Equal(t, len(a), len(b), fmt.Sprintf("a = [%s], b = [%s]", pprint(a), pprint(b)))
+	sortSliceOfByteSlices(a)
+	sortSliceOfByteSlices(b)
+	require.Equal(t, a, b)
+}
+
+func pprint(a [][]byte) string {
+	var buf bytes.Buffer
+	for i, t := range a {
+		if i != 0 {
+			buf.WriteString(", ")
+		}
+		buf.WriteString(fmt.Sprintf("%d %s", i, string(t)))
+	}
+	return buf.String()
+}
+
+func newTestSegments(t *testing.T, docs []doc.Document) (memSeg segment.MutableSegment, fstSeg segment.Segment) {
+	s := newTestMemSegment(t)
+	for _, d := range docs {
+		_, err := s.Insert(d)
+		require.NoError(t, err)
+	}
+	return s, newFSTSegment(t, s)
+}
+
+// TODO(prateek): follow up if this is a fair requirement in everyone's mind.
```

Comment:
note to self: have to track some issues in the implementation for 0 length field-values e.g. `"xyz"` => `""`. It's somewhere in fst read path.

@robskillington Do you think it's a fair requirement to require all tags to have non-empty name/value?

Reply:
Hm yeah that's probably fair, the question is though, where can/should we block this.  Potentially at the `db.WriteTagged` method we can do the validation there and bubble back an `InvalidParamsError` if so.

Reply:
Definitely. So plan for m3db would be to do that at Client/WriteTagged on client side, and in serialize.TagDecoder on server side. i.e. we'd have to wait until we're sure we need to decode the tags so that we don't do this excessively. 

For m3ninx, it'd be in `doc.Validate()` /cc @jeromefroe